kimchi/src/tests/turshi.rs

```diff
@@ -0,0 +1,178 @@
+use crate::circuits::constraints::ConstraintSystem;
+use crate::circuits::gate::CircuitGate;
+use crate::circuits::polynomials::turshi::*;
+use ark_ec::AffineCurve;
+use cairo::{helper::*, memory::CairoMemory, runner::CairoProgram};
+use mina_curves::pasta::fp::Fp as F;
+use mina_curves::pasta::pallas;
+
+type PallasField = <pallas::Affine as AffineCurve>::BaseField;
+
+/*
+fn view_witness<F: Field>(witness: &[Vec<F>; COLUMNS]) {
+    let rows = witness[0].len();
+    for i in 0..rows {
+        print!("row {}: [", i);
+        for j in 0..witness.len() {
+            print!("{} , ", witness[j][i].to_u64());
+        }
+        println!("]");
+    }
+}
+
+fn view_table<F: Field>(table: &Vec<[F; COLUMNS]>) {
+    let rows = table.len();
+    for i in 0..rows {
+        print!("row {}: [", i);
+        for j in 0..COLUMNS {
+            print!("{} , ", table[i][j].to_u64());
+        }
+        println!("]");
+    }
+}
+*/
+
+// creates a constraint system for a number of Cairo instructions
+fn create_test_consys(inirow: usize, ninstr: usize) -> ConstraintSystem<PallasField> {
+    let gates = CircuitGate::<PallasField>::create_cairo_gadget(inirow, ninstr);
+    ConstraintSystem::create(gates, vec![], oracle::pasta::fp_kimchi::params(), 0).unwrap()
+}
+
+#[test]
+fn test_cairo_cs() {
+    let instrs: Vec<i128> = vec![
+        0x400380007ffc7ffd,
+        0x482680017ffc8000,
+        1,
+        0x208b7fff7fff7ffe,
+        0x480680017fff8000,
+        10,
+        0x48307fff7fff8000,
+        0x48507fff7fff8000,
+        0x48307ffd7fff8000,
+        0x480a7ffd7fff8000,
+        0x48127ffb7fff8000,
+        0x1104800180018000,
+        -11,
+        0x48127ff87fff8000,
+        0x1104800180018000,
+        -14,
+        0x48127ff67fff8000,
+        0x1104800180018000,
+        -17,
+        0x208b7fff7fff7ffe,
+        /*41, // beginning of outputs
+        44,   // end of outputs
+        44,   // input
+        */
+    ];
+
+    let mut mem = CairoMemory::<F>::new(F::vec_to_field(&instrs));
+    // TODO(querolita): Need to know how to find out (hints)
+    mem.write(F::from(21u32), F::from(41u32)); // beginning of outputs
+    mem.write(F::from(22u32), F::from(44u32)); // end of outputs
+    mem.write(F::from(23u32), F::from(44u32)); //end of program
+    let prog = CairoProgram::new(&mut mem, 5, 24);
```

Comment:
Every time you write to memory, the length gets automatically extended. Meaning, before line 72 the length was 21 (from 0 to 20) and then in line 75 the length of memory is already 24. the argument `24` in the creator of the program corresponds to the initial allocation pointer. This means, that I think you're right and this argument could be "omitted": the initial allocation pointer should be the first empty address of the memory, so it must be 24 in this case.

Reply:
Hmm... yeah, this is interesting.  It might be cool if the memory thing was "smart" and you can just fill it up with the initial memory and the `CairoProgram::new(&mut mem)` was smart enough to figure out what to tuse for `pc` and `ap`.  Maybe a thing for future though and not this PR :-). I was just concerned if there could be a bad offset there that could impact the test case